pkg/local_object_storage/shard/tree.go

```diff
@@ -13,32 +13,44 @@ var ErrPiloramaDisabled = logicerr.New("pilorama is disabled")
 
 // TreeMove implements the pilorama.Forest interface.
 func (s *Shard) TreeMove(d pilorama.CIDDescriptor, treeID string, m *pilorama.Move) (*pilorama.LogMove, error) {
+	s.m.RLock()
+	defer s.m.RUnlock()
+
 	if s.pilorama == nil {
 		return nil, ErrPiloramaDisabled
 	}
-	if s.GetMode().ReadOnly() {
+
+	if s.info.Mode.ReadOnly() {
 		return nil, ErrReadOnlyMode
 	}
 	return s.pilorama.TreeMove(d, treeID, m)
 }
 
 // TreeAddByPath implements the pilorama.Forest interface.
 func (s *Shard) TreeAddByPath(d pilorama.CIDDescriptor, treeID string, attr string, path []string, meta []pilorama.KeyValue) ([]pilorama.LogMove, error) {
+	s.m.RLock()
+	defer s.m.RUnlock()
+
 	if s.pilorama == nil {
 		return nil, ErrPiloramaDisabled
 	}
-	if s.GetMode().ReadOnly() {
+
+	if s.info.Mode.ReadOnly() {
 		return nil, ErrReadOnlyMode
 	}
 	return s.pilorama.TreeAddByPath(d, treeID, attr, path, meta)
 }
 
 // TreeApply implements the pilorama.Forest interface.
 func (s *Shard) TreeApply(d pilorama.CIDDescriptor, treeID string, m *pilorama.Move) error {
+	s.m.RLock()
+	defer s.m.RUnlock()
+
 	if s.pilorama == nil {
```

Comment:
We can do this before locking a mutex.

Reply:
sure, just want to have the same logic as _the same as possible_ (lock in the same place (the first lines of the methods))

fixed